[Navigation] Resolve all intercept promise handlers

TingPing opened the pull request:
#### 6e7e4bcae9a41b77f6dffefd7859e54411addc5c
<pre>
[Navigation] Resolve all intercept promise handlers
<a href="https://bugs.webkit.org/show_bug.cgi?id=277340">https://bugs.webkit.org/show_bug.cgi?id=277340</a>

Reviewed by Alex Christensen.

This implements step 33.4 of <a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#update-the-navigation-api-entries-for-a-same-document-navigation">https://html.spec.whatwg.org/multipage/nav-history-apis.html#update-the-navigation-api-entries-for-a-same-document-navigation</a>

Co-authored-by: Rob Buis &lt;rbuis@igalia.com&gt;

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-and-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-multiple-times-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-multiple-times-reject-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-reject-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-same-document-history-back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigatesuccess-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/signal-abort-window-stop-after-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/back-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/forward-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-interrupted-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-interrupted-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-interrupted-within-onnavigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/traverseTo-detach-between-navigate-and-navigatesuccess-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/navigate-204-205-download-then-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/transition-finished-mark-as-handled-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/transition-realms-and-identity-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::abortOngoingNavigation):
(WebCore::createPromiseAndWrapper):
(WebCore::waitForAllPromises):
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::createPromiseAndWrapper):
* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::createPromiseAndWrapper): Deleted.

Canonical link: <a href="https://commits.webkit.org/282564@main">https://commits.webkit.org/282564@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2996b7543aaaebf6ef3c3617ec2afcf7684df951

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63522 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16119 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67543 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14130 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65642 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50566 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14410 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51152 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9774 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66591 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39807 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55010 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31837 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36490 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12381 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13002 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58000 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12702 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69239 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7469 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/12282 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58458 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7501 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55093 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58695 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14062 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6241 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/38699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39778 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40890 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39521 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -613,6 +614,50 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
     }
 }
 
+static std::pair<Ref<DOMPromise>, Ref<DeferredPromise>> createPromiseAndWrapper(Document& document)
```
At some point maybe we want to share this somewhere, I think I got it from ViewTransitions but IIRC there is even one more user.

achristensen07 replied:
Yeah, please refactor rather than copy/paste code.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -613,6 +614,50 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
+        if (promise->isSuspended())
+            return;
+
+        promise->whenSettled([=] () mutable {
```
Can you explain how the capturing works correctly in waiting for all promises to resolve? I have read about all capture modes in a C++ book some time ago but I forgot.

TingPing replied:
This copies the pointers it references, just adding refs in this case.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -717,48 +762,72 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
+            if (auto* error = jsDynamicCast<JSC::ErrorInstance*>(result))
+                errorMessage = makeString("Uncaught "_s, error->sanitizedToString(protectedScriptExecutionContext()->globalObject()));
+            auto exception = Exception(ExceptionCode::UnknownError, errorMessage);
+
```
A FIXME would be nice for the line info, I did this to progress on some tests.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -613,6 +614,50 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
+}
+
+// https://webidl.spec.whatwg.org/#wait-for-all
+static void waitForAllPromises(Vector<RefPtr<DOMPromise>>& promises, std::function<void()> fulfilledCallback, std::function<void(JSC::JSValue)> rejectionCallback)
```
Can we make promises a const Vector& or a Vector&&?  As it stands now, it looks like an out param, which isn't great.

Also, can we use WTF::Function instead of std::function?

TingPing replied:
> Also, can we use WTF::Function instead of std::function?

In this case since the function is called later it doesn't work out. A ref to `WTF::Function` doesn't work as it goes out of scope and `WTF::Function` is not copyable unlike `std::function`.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -613,6 +614,50 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
+// https://webidl.spec.whatwg.org/#wait-for-all
+static void waitForAllPromises(Vector<RefPtr<DOMPromise>>& promises, std::function<void()> fulfilledCallback, std::function<void(JSC::JSValue)> rejectionCallback)
+{
+    auto remainingPromises = std::make_shared<size_t>(promises.size());
```
WebKit usually uses RefPtr and RefCounted instead of std::shared_ptr.  Can we make a structure of the state?  Can we use CallbackAggregator instead?

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -613,6 +614,50 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
+
+    for (const auto& promise : promises) {
+        // At any point between promises the frame could have been detached.
+        if (promise->isSuspended())
```
So if all promises are suspended we will never call either callback.  Is that what we want?  Can they become unsuspended?

TingPing replied:
I'm not certain its possible and I'm not sure what to do in that case.

It may not be possible, like `whenSettled`'s callback probably can't be called from JS when JS is already suspended. These codepaths have only been hit when the frame is actively destroyed during the NavigateEvent.

TingPing replied:
I did try calling rejection in this case also but no test case covered it. I could maybe write one expecting an error after detach to see if that edge-case matters.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -717,48 +762,72 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
-            }
-        } else {
-            // FIXME: and the following failure steps given reason rejectionReason:
+        }, [=, this](JSC::JSValue result) {
```
How do we know that this has not been destroyed when the lambda is called?

TingPing replied:
Added WeakRef.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -717,48 +762,72 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
+                transition->rejectPromise(exception);
+
+            if (apiMethodTracker) {
+                apiMethodTracker->finishedPromise->rejectWithCallback([result] (auto&) {
```
Is it safe to capture result here?  I'm honestly not sure.

TingPing replied:
I think it's fine, copying a JSValue. This callback will only be active when JS is too.

achristensen07 replied:
I'm not sure it is fine.  JS being active is different than whether result has been garbage collected or not.  We may need to protect and unprotect it or something.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -613,6 +614,51 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
+static void waitForAllPromises(const Vector<RefPtr<DOMPromise>>& promises, Function<void()>&& fulfilledCallback, Function<void(JSC::JSValue)>&& rejectionCallback)
+{
+    Ref awaitingData = adoptRef(*new AwaitingPromiseData);
+    awaitingData->remainingPromises = promises.size();
```
Can these instead be passed to the constructor with an initializer list?
Ref awaitingData = adoptRef(*new AwaitingPromiseData {
    WTFMove(fulfilledCallback),
    WTFMove(rejectionCallback),
    promises.size(),
    false
});

That way, if we forget something in the future it'll be a compiler error instead of something we'll have to manually notice.

TingPing replied:
So I believe this was completely safe, as the callback is called immediately.

However I found a method that avoids the callback entirely which seems cleaner.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -613,6 +614,51 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
+            return;
+
+        promise->whenSettled([=] () mutable {
+            if (promise->isSuspended())
```
If these early returns are hit, the remainingPromises check won't be correct, will it?  Will that result in promises being never fulfilled or rejected?

TingPing replied:
Yes, but I can't see any other way to handle this differently.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -717,48 +763,73 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
-            // If a new event has been dispatched in our event handler then we were aborted above.
-            if (m_ongoingNavigateEvent != event.ptr())
-                return false;
+        waitForAllPromises(promiseList, [=, this, weakThis = WeakPtr { *this }]() mutable {
```
Can we remove the = and explicitly list everything that is needed to be captured?  That will prevent us from accidentally implicitly capturing a raw pointer.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -717,48 +763,73 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
-            if (m_ongoingNavigateEvent != event.ptr())
-                return false;
+        waitForAllPromises(promiseList, [=, this, weakThis = WeakPtr { *this }]() mutable {
+            if (!weakThis || abortController->signal().aborted() || !document->isFullyActive() || !m_ongoingNavigateEvent)
```
Can we make a RefPtr strongThis = weakThis.get()?  That'll prevent a callback destroying the Navigation object somehow from hitting a UAF with the raw this pointer.  Same below.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -613,6 +614,51 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
+        if (promise->isSuspended())
+            return;
+
+        promise->whenSettled([=] () mutable {
```
Could we instead explicitly write out what is being captured?  We don't want to implicitly capture the this pointer without a lifetime check, for example.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -613,6 +614,56 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
+        , rejectionCallback(WTFMove(rejectionCallback))
+        , remainingPromises(remainingPromises)
+    {
+    };
```
nit: extra semicolon not needed.

TingPing committed:
[Navigation] Resolve all intercept promise handlers
https://bugs.webkit.org/show_bug.cgi?id=277340

Reviewed by Alex Christensen.

This implements step 33.4 of https://html.spec.whatwg.org/multipage/nav-history-apis.html#update-the-navigation-api-entries-for-a-same-document-navigation

Co-authored-by: Rob Buis <rbuis@igalia.com>

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-and-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-multiple-times-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-multiple-times-reject-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-reject-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-same-document-history-back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigatesuccess-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/signal-abort-window-stop-after-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/back-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/forward-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-interrupted-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-interrupted-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-interrupted-within-onnavigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/traverseTo-detach-between-navigate-and-navigatesuccess-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/navigate-204-205-download-then-same-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/transition-finished-mark-as-handled-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/transition-realms-and-identity-expected.txt:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::abortOngoingNavigation):
(WebCore::createPromiseAndWrapper):
(WebCore::waitForAllPromises):
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::createPromiseAndWrapper):
* Source/WebCore/bindings/js/JSDOMPromiseDeferred.h:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::createPromiseAndWrapper): Deleted.

Canonical link: https://commits.webkit.org/282564@main
